### ISSUE TITLE
KAGGLE / Computer Vision Tutorials / Custom Convets & Data Augmentation

### ISSUE THREAD

**Issue by sallyy1** [sallyy1/Data-Analysis#3]
### [**Convolutional Blocks**]
It does this by passing them through long chains of convolutional blocks which perform this extraction.
![image](https://user-images.githubusercontent.com/70500841/97669446-780f1300-1ac7-11eb-8348-ada6261de35c.png)
These convolutional blocks are stacks of **Conv2D** and **MaxPool2D** layers, whose role in feature extraction we learned about in the last few lessons.
![image](https://user-images.githubusercontent.com/70500841/97669472-86f5c580-1ac7-11eb-9ec5-8d433afb15fd.png)

Each block represents a round of extraction, and by composing these blocks the convnet can combine and recombine the features produced, growing them and shaping them to better fit the problem at hand. The deep structure of modern convnets is what allows this sophisticated feature engineering and has been largely responsible for their superior performance.

### [**Example - Design a Convnet**]

**Step 1 - Load Data**
This hidden cell loads the data.

**Step 2 - Define Model**
Here is a diagram of the model we'll use:

![image](https://user-images.githubusercontent.com/70500841/97669555-b573a080-1ac7-11eb-82a1-b628997ba265.png)

Now we'll define the model. See how our model consists of three blocks of **Conv2D** and **MaxPool2D** layers (the base) followed by a head of **Dense** layers. We can translate this diagram more or less directly into a Keras **Sequential** model just by filling in the appropriate parameters.

```
from tensorflow import keras
from tensorflow.keras import layers

model = keras.Sequential([

    # First Convolutional Block
    layers.Conv2D(filters=32, kernel_size=5, activation="relu", padding='same',
                  # give the input dimensions in the first layer
                  # [height, width, color channels(RGB)]
                  input_shape=[128, 128, 3]),
    layers.MaxPool2D(),

    # Second Convolutional Block
    layers.Conv2D(filters=64, kernel_size=3, activation="relu", padding='same'),
    layers.MaxPool2D(),

    # Third Convolutional Block
    layers.Conv2D(filters=128, kernel_size=3, activation="relu", padding='same'),
    layers.MaxPool2D(),

    # Classifier Head
    layers.Flatten(),
    layers.Dense(units=6, activation="relu"),
    layers.Dense(units=1, activation="sigmoid"),
])
model.summary()
```

**Step 3 - Train**

```
model.compile(
    optimizer=tf.keras.optimizers.Adam(epsilon=0.01),
    loss='binary_crossentropy',
    metrics=['binary_accuracy']
)

history = model.fit(
    ds_train,
    validation_data=ds_valid,
    epochs=40,
)
```

```
import pandas as pd

history_frame = pd.DataFrame(history.history)
history_frame.loc[:, ['loss', 'val_loss']].plot()
history_frame.loc[:, ['binary_accuracy', 'val_binary_accuracy']].plot();
```

-------------
### [**Using Data Augmentation**]
-So, if we augment our training data with flipped images, our classifier will learn that "left or right" is a difference it should ignore.
-rotating the image, adjusting the color or contrast, warping the image, or many other things, usually applied in combination.

### [**Example - Training with Data Augmentation**]
Keras lets you augment your data in two ways. The first way is to include it in the data pipeline with a function like ImageDataGenerator. The second way is to include it in the model definition by using Keras's **preprocessing layers**. This is the approach that we'll take. The primary advantage for us is that the image transformations will be computed on the GPU instead of the CPU, potentially speeding up training.

**Step 1 - Load Data**

**Step 2 - Define Model**
To illustrate the effect of augmentation, we'll just add a couple of simple transformations to the model from Tutorial 1.

```
from tensorflow import keras
from tensorflow.keras import layers
# these are a new feature in TF 2.2
from tensorflow.keras.layers.experimental import preprocessing


pretrained_base = tf.keras.models.load_model(
    '../input/cv-course-models/cv-course-models/vgg16-pretrained-base',
)
pretrained_base.trainable = False

model = keras.Sequential([
    # Preprocessing
    preprocessing.RandomFlip('horizontal'), # flip left-to-right
    preprocessing.RandomContrast(0.5), # contrast change by up to 50%
    # Base
    pretrained_base,
    # Head
    layers.Flatten(),
    layers.Dense(6, activation='relu'),
    layers.Dense(1, activation='sigmoid'),
])
```
**Step 3 - Train and Evaluate**

```
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['binary_accuracy'],
)

history = model.fit(
    ds_train,
    validation_data=ds_valid,
    epochs=30,
)
```
```
import pandas as pd

history_frame = pd.DataFrame(history.history)

history_frame.loc[:, ['loss', 'val_loss']].plot()
history_frame.loc[:, ['binary_accuracy', 'val_binary_accuracy']].plot();
```
